### PR TITLE
update new rule

### PR DESCRIPTION
**Pull request by mercury233**

### CODE REVIEW

File: c17745969.lua [hunk 1]
```diff
@@ -34,25 +34,22 @@ function c17745969.spfilter1(c,e,tp,loc)
...
+			or c:IsLocation(LOCATION_EXTRA) and Duel.GetLocationCountFromEx(tp,tp,nil,c)>0)
 end
 function c17745969.sptg(e,tp,eg,ep,ev,re,r,rp,chk,chkc)
-	local loc=0
-	if Duel.GetLocationCount(tp,LOCATION_MZONE)>0 then loc=loc+LOCATION_DECK end
-	if Duel.GetLocationCountFromEx(tp)>0 then loc=loc+LOCATION_EXTRA end
+	local loc=LOCATION_DECK+LOCATION_EXTRA
 	if chkc then return chkc:IsControler(1-tp) and chkc:IsLocation(LOCATION_MZONE) and c17745969.spfilter1(chkc,e,tp,loc) end
-	if chk==0 then return loc~=0 and Duel.IsExistingTarget(c17745969.spfilter1,tp,0,LOCATION_MZONE,1,nil,e,tp,loc) end
+	if chk==0 then return Duel.IsExistingTarget(c17745969.spfilter1,tp,0,LOCATION_MZONE,1,nil,e,tp,loc) end
```
Comment (DailyShana) [inline c17745969.lua]:
```suggestion
	if chk==0 then return Duel.IsExistingTarget(c17745969.spfilter1,tp,0,LOCATION_MZONE,1,nil,e,tp,LOCATION_DECK+LOCATION_EXTRA) end
```

File: c29085954.lua [hunk 2]
```diff
@@ -27,18 +27,17 @@ function c29085954.filter(c,e,tp,mc)
 	local no=m.xyz_number
 	return no and no>=1 and no<=99 and c:IsSetCard(0x48)
 		and mc:IsCanBeXyzMaterial(c)
-		and c:IsCanBeSpecialSummoned(e,SUMMON_TYPE_XYZ,tp,false,false)
+		and c:IsCanBeSpecialSummoned(e,SUMMON_TYPE_XYZ,tp,false,false) and Duel.GetLocationCountFromEx(tp,tp,mc,c)>0
 end
 function c29085954.sptg(e,tp,eg,ep,ev,re,r,rp,chk)
-	if chk==0 then return Duel.GetLocationCountFromEx(tp,tp,e:GetHandler())>0
```
Comment (DailyShana) [inline c29085954.lua]: no location check in target?

File: c1412158.lua [hunk 3]
```diff
@@ -34,18 +34,18 @@ end
...
-		and Duel.IsExistingMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,1,nil,e,tp) end
+		and Duel.IsExistingMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,1,nil,e,tp,e:GetHandler()) end
 	Duel.SetOperationInfo(0,CATEGORY_TODECK,e:GetHandler(),1,0,0)
 	Duel.SetOperationInfo(0,CATEGORY_SPECIAL_SUMMON,nil,1,tp,LOCATION_EXTRA)
 end
 function c1412158.spop(e,tp,eg,ep,ev,re,r,rp)
 	local c=e:GetHandler()
-	local tc=Duel.GetFirstMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,nil,e,tp)
+	local tc=Duel.GetFirstMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,nil,e,tp,nil)
 	if tc and c:IsRelateToEffect(e) and c:IsFaceup() and Duel.SendtoDeck(c,nil,2,REASON_EFFECT)~=0 then
```
Comment (DailyShana) [inline c1412158.lua]: e:GetHandler() is sent to extra in chain solving so do not need to such change

File: c51476410.lua [hunk 4]
```diff
@@ -38,8 +38,9 @@ end
 function c51476410.spfilter2(c,e,tp,m,f,chkf)
 	return (not f or f(c)) and c:IsCanBeSpecialSummoned(e,SUMMON_TYPE_FUSION,tp,false,true) and c:CheckFusionMaterial(m,nil,chkf)
 end
-function c51476410.spfilter3(c,e,tp,chkf)
+function c51476410.spfilter3(c,e,tp,chkf,rc)
 	if not c:IsType(TYPE_FUSION) or not c:IsAbleToExtra() then return false end
+	if not Duel.GetLocationCountFromEx(tp,tp,rc,c)>0 then return false end
```
Comment (DailyShana) [inline c51476410.lua]: use <= directly

File: c1412158.lua (hunk 3 above)
Comment (mercury233) [reply c1412158.lua]: write it explicitly

File: c1412158.lua [hunk 5]
```diff
@@ -34,23 +34,19 @@ end
...
 end
 function c1412158.spop(e,tp,eg,ep,ev,re,r,rp)
 	local c=e:GetHandler()
-	if Duel.GetLocationCountFromEx(tp,tp,c)<=0 then return end
-	if c:IsRelateToEffect(e) and c:IsFaceup() and Duel.SendtoDeck(c,nil,2,REASON_EFFECT)~=0 then
-		local tc=Duel.GetFirstMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,nil,e,tp)
-		if tc then
-			Duel.SpecialSummon(tc,0,tp,tp,false,false,POS_FACEUP)
-		end
+	local tc=Duel.GetFirstMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,nil,e,tp,nil)
```
Comment (DailyShana) [inline c1412158.lua]: so it should be
```suggestion
	local tc=Duel.GetFirstMatchingCard(c1412158.spfilter,tp,LOCATION_EXTRA,0,nil,e,tp,c)
```

File: c75457624.lua [hunk 6]
```diff
@@ -14,34 +14,33 @@ end
...
 	Duel.Remove(g1,POS_FACEUP,REASON_COST)
 end
-function c75457624.filter(c,e,tp)
-	return c:IsSetCard(0xb5) and c:IsCanBeSpecialSummoned(e,0,tp,true,false)
+function c75457624.filter(c,e,tp,mg)
+	return c:IsSetCard(0xb5) and c:IsCanBeSpecialSummoned(e,0,tp,true,false) and Duel.GetLocationCountFromEx(tp,tp,mg,c)>0
 end
 function c75457624.target(e,tp,eg,ep,ev,re,r,rp,chk)
-	if chk==0 then return Duel.IsExistingMatchingCard(c75457624.filter,tp,LOCATION_EXTRA,0,1,nil,e,tp) end
+	if chk==0 then return Duel.IsExistingMatchingCard(c75457624.filter,tp,LOCATION_EXTRA,0,1,nil,e,tp,nil) end
```
Comment (DailyShana) [inline c75457624.lua]: if no empty monster zones, this check seems cannot pass

**Comment by mercury233** [comment]
all cards should be done

**Pull request merged**